<h1 id="advice-from-chat-with-greta-bocedi">Advice from chat with Greta Bocedi</h1>
<h3 id="general">General:</h3>
<ul>
<li>Start very simple and add complexity gradually.</li>
<li>Keep dispersal ability the same for diploids and polyploids, (simple at first just like a king) so focus is on gigas-effects.</li>
<li>Test the gigas-effects one by one.</li>
<li>Test masking of deleterious mutation and adaptation to niches separately.</li>
</ul>
<blockquote>
<p>If habitats are numbered 1-10, and genes (also numbered 1-10) code for adaptation, then polyploids will have access to more matching genes that could be made use of (this may be too simplistic? as polyploids can also have novel adaptations - see genetics below). Also start simpler with binary habitats - add more if needed.</p>
</blockquote>
<h3 id="spatial-resolution">Spatial resolution:</h3>
<p>As it’s not based on a real geographical location, resolution will not bias my dispersal ranges or population growth etc. There is also no need to use a dispersal kernel for this model (So, I can have my landscape in squares, and my dispersal also in squares), though there are kernel models for pollen and seeds that could be used if wanted.</p>
<h2 id="questions-remaining">Questions remaining</h2>
<h3 id="can-i-make-the-improved-growth-rateother-gigas-effects-have-a-genetic-basis">Can I make the improved growth rate/other gigas effects have a genetic basis?</h3>
<p>My thoughts are that I would like the genetic complexity to be fairly detailed as it’s for the genetics society. This will make the evolution and spread of polyploid lineages through generations more accurate. And it will give a number of mechanisms that could be tested separately to see which have the largest impact on results.</p>
<h3 id="do-gigas-effects-affect-dispersal-ability">Do gigas-effects affect dispersal ability?</h3>
<p>Perhaps by:</p>
<ul>
<li>increasing seed wing size to bias wind dispersal range?</li>
<li>fruit size to bias frugivore dispersal?</li>
<li>inflorescence size to bias pollinators?</li>
</ul>
<p><strong>Should a genetic basis for these traits be included?</strong></p>
<h2 id="points-from-dinner">Points from dinner</h2>
<p>Brad talked about some kind of infinite allele model that we will use.</p>
<blockquote>
<p>“Paul Joyce and the infinite alleles model” added to Mendeley (READ).</p>
</blockquote>
<h2 id="genetically-explicit-details">Genetically Explicit Details</h2>
<p><a href="https://www.youtube.com/watch?v=JVro1y24IBA">See Wendell’s talk</a><br>
Crosstalk between genomes that fully/partially converts homologous genes. This mechanism can delete deleterious mutation as well as create brand new adaptation/alleles that neither progenitor possesed.</p>
<ul>
<li>When does this occur?</li>
</ul>
<p>Genome downsizing. Sudden burst of gene loss (10x that of diploids) for neopolyploids that reduces gradually through generations.</p>
<ul>
<li>until when? diploid again?</li>
</ul>
<p>Biased expression of one genome/genes</p>
<ul>
<li>perhaps influenced by the environmental conditions encountered?</li>
<li>Not sure if this needs to be modelled explicitly or if it will just emerge?</li>
</ul>
<h3 id="cycle-according-to-wendell">Cycle according to Wendell</h3>
<ol>
<li>Whole genome doubling.</li>
<li>Biased expression:
<ul>
<li>Neofunctionalization/Subfunctionalization/Gene loss/Gene transfer/etc</li>
<li>Gene expression tends to bias to one of the two parents.</li>
</ul>
</li>
<li>Biased fractionation:
<ul>
<li>Different kinds of genes are more likely to survive polyploid events.</li>
<li>One genome usually experiences more gene loss than the other.</li>
<li>Difference in size between progenitor genomes can set direction of bias (in which direction?)</li>
</ul>
</li>
<li>Massive genome downsizing:
<ul>
<li>Chromosome rearrangement and number reduction</li>
<li>Gene loss in one genome facilitates biased expression and growth of the other.</li>
</ul>
</li>
<li>Some time later, repeat.</li>
</ol>

