<h1 id="polyploidy">Polyploidy</h1>
<p>Seminar</p>
<ul>
<li><a href="https://www.eeob.iastate.edu/people/jonathan-wendel">Dr. Jonathan F Wendel</a></li>
<li><a href="https://www.youtube.com/watch?v=JVro1y24IBA">similar seminar on youtube</a></li>
</ul>
<h2 id="gossypium-cotton-genus">Gossypium (cotton genus)</h2>
<p>All cotton species hire ants to defend. except the hawaiian species.<br>
Natural cotton has barely any fibrous/hairy seeds. May wild forms have reduced hair and reduced numbers of seeds in comparison to the typical domestic cotton plant we are used to seeing.<br>
Human breeding selection pressures have lead to the extremely fibrous, high yielding plants we see today.<br>
Domestication of cotton began in the New World/Mexico around the same time as the Spanish Armada (when pyramids were being built. really!?)</p>
<p>Common ancestor originated 5-10 mya<br>
Now there are species in the America, Peru, Hawaii, Galapagos, Africa and Australia. The continents haven’t changed much in the last 10 MY so we know transoceanic dispersal has occurred.</p>
<p>The most important was Africa (A) to central America (D).<br>
A + D hybridisation the first allopolyploid Gossypium (~2 MYA).<br>
Both diploid species had been in isolation, in totally different ecological settings for 5-10 MY <strong>and yet were still able to hybridise</strong>.</p>
<blockquote>
<p>most recent duplication event. really two paleo 72 polyploids hybridised, not two diploids.</p>
</blockquote>
<p>Both genomes (AADD) exist in the new nucleus, but only one of the two parental cytoplasms is retained (in this case A cytoplasms and mitochondria).</p>
<p>These differentiated to 6 polyploid AADD species which we use agriculturally today.<br>
Allopolyploids have phenotypes which are transgressive to their diploid progenitors.</p>
<h2 id="polyploidy-1">polyploidy</h2>
<ul>
<li><strong>homeologs</strong> = 2 copies of chromosomes that are similar to each other or</li>
<li><strong>homeologous genes</strong> = 2 duplicated genes, 1 on each chromosome</li>
<li>biased expression = genes from one genome is expressed more than the other</li>
<li>unbiased expression = genes are expressed equally form each genome</li>
<li>biased fractionation = different kinds of genes are more likely to survive polyploid events. And one genome usually experiences more gene loss than the other.</li>
<li>neofunctionalization = find a new job (HOT TOPIC) adaptation. See ~45mins in youtube vid for reading.</li>
<li>subfunctionalization = partition the old job between the two duplicate copies.</li>
</ul>
<p>Biased fractionation can occur at every level in every tissue and is very common. Gene expression tends to bias to one of the two parents. A phenomenon that everyone is describing that we don’t fully understand.</p>
<h3 id="genomicsdna">genomics/DNA</h3>
<p>We see LOTS of non-mendelian behaviours.<br>
We do not expect to see gene transfer in strict bivalent pairing, but we do.<br>
There is <strong>gene conversion</strong>. In fact it’s common.<br>
The homologous genes talk to each other and can overwrite each other, partially or fully.</p>
<ul>
<li>diversity reducing / fully overwriting extra duplicated alleles with one copy or the other.</li>
<li>diversity increasing / partially overwriting to create completely new alleles.<br>
Therefore, this gives the ability to correct deleterious mutation (current area of research).</li>
<li>hard to get rid of deleterious mutations in centromeres especially</li>
<li>polyploidy may provide an escape route for that.</li>
</ul>
<blockquote>
<p>The Cytonuclear Dimension of Allopolyploid Evolution: An Example from Cotton Using Rubisco (READ)<br>
Added to Mendeley: <a href="https://academic.oup.com/mbe/article/29/10/3023/1027173">Article</a></p>
</blockquote>
<p>There is <strong>gene loss</strong>, as well as unequal rates and intergenomic transfer.</p>
<p>Larger genomes accumulate more diversity and have more gene loss.<br>
Gene loss 10 x more frequent in allopolyploids than diploids.<br>
Most loss early on after the duplication event.</p>
<h3 id="expressionrna">expression/RNA</h3>
<p>Biased expression of duplicated genes.<br>
Reciprocal silencing of alternative homoeologs.</p>
<blockquote>
<p>There is no single organism that does not have this bias. (CHALLENGE)</p>
</blockquote>
<p>Also happens at the genome scale.<br>
One genome will be used more than the other.<br>
Just as one gene will be expressed more than its duplicate.</p>
<p>Allopolyploid cotton (sp?) really has 144 genomes phylogenetically according to all polyploid events in it’s lineage.<br>
This would give a much bigger number of chromosomes than what we actually see.<br>
Massive gene loss / genome downsizing has occurred - to get rid of junk.<br>
This is true for all polyploids.</p>
<h4 id="transcription-factors-incredibly-retained.">Transcription factors incredibly retained.</h4>
<p>A plant genome can commonly have hundreds of mib transcription factors (important in metabolic pathways). Why do they have so many when animals typically have around 3?<br>
There doesn’t need to be a novel propagation mechanism.<br>
They are just selectively retained during polyploidy events.<br>
<strong>Genes of different types are subject to different selection pressures.</strong></p>
<h4 id="difference-in-genome-size-between-progenitors-contributes-to-direction-of-biased-fractionation">Difference in genome size between progenitors contributes to direction of biased fractionation</h4>
<p>How different in size are the two genomes at point of hybridisation?</p>
<ul>
<li>this can set biased fractionation going - in which direction?</li>
<li>then the gene loss of one genome seems to facilitate the over expression and growth of the alternate genome.</li>
</ul>
<hr>
<h2 id="cycle">Cycle</h2>
<ol>
<li>Whole genome doubling</li>
<li>Biased expression (Neofunctionalization/Subfunctionalization/Gene loss/Gene transfer/etc)</li>
<li>Biased fractionation</li>
<li>Massive genome downsizing</li>
<li>Chromosome rearrangement and number reduction</li>
<li>Some time later, repeat.</li>
</ol>
<hr>
<h2 id="why-is-polyploidy-common-in-some-groups-and-rare-in-others">why is polyploidy common in some groups and rare in others?</h2>
<p>why is it so rare in legumes?<br>
but so common in particular grasses, or rosacea.</p>
<p><strong>this question is one we’ve made almost no progress in.</strong></p>
<hr>
<p>Marine is talking in 2 weeks about polyploid snails<br>
Thought to be less important in fungi, but not much known about that yet.</p>

